compressable:Zstd comp support

Athishpranav2003 opened the pull request:
<!--
Thank you for contributing to Fluentd!
Your commits need to follow DCO: https://probot.github.io/apps/dco/
And please provide the following information to help us make the most of your pull request:
-->

**Which issue(s) this PR fixes**: 
Fixes #4162 

**What this PR does / why we need it**: 
Adds new compression method support to handle messages
**Docs Changes**:
TODO
**Release Note**: 
N/A

Athishpranav2003 commented:
@daipom @ashie need some comments on this since its a very new compression method. I tried to make sure the existing support is not broken and added this feature additionally. Need to perform some more work but wanted to get some comments from you guys

daipom commented:
Thanks! I will see this this weekend!

Athishpranav2003 commented:
@daipom did u get chance to look at this?

Athishpranav2003 commented:
Its fine i guess,
Even i was very busy in last 2 weeks :+1:

daipom commented:
@Athishpranav2003 
I'm sorry for my late response. :cry:

I have confirmed the entire direction!
It's great! Thanks for starting this enhancement!

Sorry I haven't made time to see the detailed implementation, such as `Compressable` module,
but the overall design of Fluentd would be essential for us now.
So, now, I comment on the overall direction of further modification.

It looks basically good as `in_forward` support!
All that is left is to support Output/Buffer/Chunk logic for output plugins. (Currently, these logic assumes `gzip` only)
To do so, it would be a good idea to support `out_forward` first.

The core logic would be the following.

* `EventStream#to_compressed_msgpack_stream`
  * https://github.com/fluent/fluentd/blob/78a7972bfe6b421f08472701f04f00515ed24bee/lib/fluent/event.rb#L65-L68
* `Output#generate_format_proc`
  * https://github.com/fluent/fluentd/blob/78a7972bfe6b421f08472701f04f00515ed24bee/lib/fluent/plugin/output.rb#L1016-L1027
* `Chunk::Decompressable`
  * https://github.com/fluent/fluentd/blob/78a7972bfe6b421f08472701f04f00515ed24bee/lib/fluent/plugin/buffer/chunk.rb#L184-L235

It is complicated, but the compression of Buffer and Chunk and the flush process of output plugins are closely related.
The following existing implementations may be helpful for how output plugins behaves:

* https://github.com/fluent/fluentd/blob/78a7972bfe6b421f08472701f04f00515ed24bee/lib/fluent/plugin/out_forward.rb#L253-L259
* https://github.com/fluent/fluentd/blob/78a7972bfe6b421f08472701f04f00515ed24bee/lib/fluent/plugin/out_forward.rb#L664-L668
* https://github.com/fluent/fluentd/blob/78a7972bfe6b421f08472701f04f00515ed24bee/lib/fluent/plugin/out_file.rb#L214-L225
* https://github.com/fluent/fluentd/blob/78a7972bfe6b421f08472701f04f00515ed24bee/lib/fluent/plugin/out_file.rb#L245-L268

Athishpranav2003 commented:
@daipom thanks for the review.

Yah, this will kick a lot more changes in the overall events pipeline.
For now i have just added the compression support as a module in `Compressible`. And used this in `in-forward`. I will try to look at Buffer as well in meantime(might need more time since i got busy with a project). I guess maybe splitting it into multiple PRs will be easy since i can finish small chunks one by one(seperate functionality) and easy to review

daipom commented:
@Athishpranav2003 
> can you please check the change in the Compression module as well

Sure! I will review the `Compressable` module!

> Once if we merge this we can proceed with the other development around this

To judge if we can merge this only with the support of the module and `in_forward`, we need to reach an agreement on updating [Forward Protocol Specification](https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1.5).

It would be necessary to update `CompressedPackedForward Mode`.
By allowing `zstd` as the value for the key `compressed`, it would be possible to add Zstd support while keeping compatibility, but I'm not sure yet.

Athishpranav2003 commented:
Sure @daipom 

If it's needed then I can split the PR into 2 where this will only focus on `zstd` and other one for `in-forward`

Athishpranav2003 commented:
@daipom how to proceed in this PR?

daipom commented:
Sorry, please give me a few more days 😢

daipom commented on lib/fluent/event.rb:
```diff
@@ -203,7 +203,7 @@ class MessagePackEventStream < EventStream
 
     # Keep cached_unpacker argument for existing plugins
-    def initialize(data, cached_unpacker = nil, size = 0, unpacked_times: nil, unpacked_records: nil)
+    def initialize(data, cached_unpacker = nil, size = 0, unpacked_times: nil, unpacked_records: nil, compress: nil)
```
```suggestion
    def initialize(data, cached_unpacker = nil, size = 0, unpacked_times: nil, unpacked_records: nil)
```

daipom commented on lib/fluent/event.rb:
```diff
@@ -268,10 +268,11 @@ def to_msgpack_stream(time_int: false, packer: nil)
 
   class CompressedMessagePackEventStream < MessagePackEventStream
-    def initialize(data, cached_unpacker = nil, size = 0, unpacked_times: nil, unpacked_records: nil)
+    def initialize(data, cached_unpacker = nil, size = 0, unpacked_times: nil, unpacked_records: nil, compress: :gzip)
       super
```
```suggestion
    def initialize(data, cached_unpacker = nil, size = 0, unpacked_times: nil, unpacked_records: nil, compress: :gzip)
      super(data, cached_unpacker, size, unpacked_times: unpacked_times, unpacked_records: unpacked_records)
```

daipom commented on lib/fluent/plugin/compressable.rb:
```diff
@@ -16,75 +16,101 @@
-      def compress(data, **kwargs)
+      def compress(data, type: :gzip, **kwargs)
         output_io = kwargs[:output_io]
+        writer = nil
```
This line would be unnecessary.

```suggestion
```

daipom commented on lib/fluent/plugin/compressable.rb:
```diff
@@ -16,75 +16,101 @@
       private
 
-      def string_decompress(compressed_data)
+      def string_decompress(compressed_data, type = :gzip)
```
How about splitting the logic for simplicity?

```rb
      def string_decompress(compressed_data, type = :gzip)
        if type == :gzip
          string_decompress_gzip(compressed_data)
        elsif type == :zstd
          string_decompress_zstd(compressed_data)
        else
          raise ArgumentError, "Unknown compression type: #{type}"
        end
      end
```

daipom commented on lib/fluent/plugin/compressable.rb:
```diff
@@ -16,75 +16,101 @@
       end
 
-      def io_decompress(input, output)
+      def io_decompress(input, output, type = :gzip)
```
ditto

daipom commented on lib/fluent/plugin/in_forward.rb:
```diff
@@ -309,8 +309,7 @@ def on_message(msg, chunk_size, conn)
         size = (option && option['size']) || 0
-        es_class = (option && option['compressed'] == 'gzip') ? Fluent::CompressedMessagePackEventStream : Fluent::MessagePackEventStream
-        es = es_class.new(entries, nil, size.to_i)
+        es = (option && option['compressed']!=nil && option['compressed']!="text") ? Fluent::CompressedMessagePackEventStream.new(entries, nil, size.to_i, compress: option['compressed'].to_sym) : Fluent::MessagePackEventStream.new(entries, nil, size.to_i)
```
How about this for simplicity?

```rb
        option = msg[2] || {}

        size = option['size'] || 0
        if option['compressed'] == 'gzip' || option['compressed'] == 'zstd'
          es = Fluent::CompressedMessagePackEventStream.new(entries, nil, size.to_i, compress: option['compressed'].to_sym)
        else
          es = Fluent::MessagePackEventStream.new(entries, nil, size.to_i)
        end
```

daipom reviewed (commented):
@Athishpranav2003 
Sorry for my late response :cry: 
I have commented on some points, but basically it looks good! Thanks!

About our future policies, I think it is better not to separate the PRs, as these changes need to be considered in combination with updating [Forward-Protocol](https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1.5#packedforward-mode).

How about supporting the `out_forward` by implementing Output/Buffer/Chunk logic in this PR?

My concern is whether we can support Buffer/Chunk zstd compression.
To support it, the compressed data must be able to be concatenated.
gzip allows it.
It appears that we can concat zstd files without any problems, but I wonder if this is an officially supported specification of zstd and `zstd-ruby` gem.

```console
$ echo "Hello world" > 1
$ echo "Hello Fluentd" > 2
$ zstd -f 1
$ zstd -f 2
$ cat 1.zst 2.zst > 3.zst
$ zstd -d 3.zst
$ cat 3.zst
Hello world
Hello Fluentd
```

Athishpranav2003 commented:
```
bundle exec rake test TEST=test/plugin/test_compressable.rb                      2856ms  Sun 03 Nov 2024 11:29:54 AM IST
/usr/bin/ruby -w -I"lib:test" -Eascii-8bit:ascii-8bit /usr/share/gems/gems/rake-13.2.1/lib/rake/rake_test_loader.rb "test/plugin/test_compressable.rb" 
/usr/share/gems/gems/bundler-2.5.16/lib/bundler/rubygems_ext.rb:250: warning: method redefined; discarding old encode_with
/usr/local/share/ruby/site_ruby/rubygems/dependency.rb:341: warning: previous definition of encode_with was here
/usr/share/ruby/bundled_gems.rb:75:in `require': libruby.so.3.2: cannot open shared object file: No such file or directory - /usr/lib64/gems/ruby/yajl-ruby-1.4.3/yajl/yajl.so (LoadError)
	from /usr/share/ruby/bundled_gems.rb:75:in `block (2 levels) in replace_require'
	from /usr/share/gems/gems/yajl-ruby-1.4.3/lib/yajl.rb:1:in `<top (required)>'
	from /usr/share/ruby/bundled_gems.rb:75:in `require'
	from /usr/share/ruby/bundled_gems.rb:75:in `block (2 levels) in replace_require'
	from /home/aggressive_racer1/projects/fluentd/lib/fluent/config/literal_parser.rb:20:in `<top (required)>'
	from /usr/share/ruby/bundled_gems.rb:75:in `require'
	from /usr/share/ruby/bundled_gems.rb:75:in `block (2 levels) in replace_require'
	from /home/aggressive_racer1/projects/fluentd/lib/fluent/config/element.rb:18:in `<top (required)>'
	from /usr/share/ruby/bundled_gems.rb:75:in `require'
	from /usr/share/ruby/bundled_gems.rb:75:in `block (2 levels) in replace_require'
	from /home/aggressive_racer1/projects/fluentd/test/helper.rb:42:in `<top (required)>'
	from /home/aggressive_racer1/projects/fluentd/test/plugin/test_compressable.rb:1:in `require_relative'
	from /home/aggressive_racer1/projects/fluentd/test/plugin/test_compressable.rb:1:in `<top (required)>'
	from /usr/share/ruby/bundled_gems.rb:75:in `require'
	from /usr/share/ruby/bundled_gems.rb:75:in `block (2 levels) in replace_require'
	from /usr/share/gems/gems/rake-13.2.1/lib/rake/rake_test_loader.rb:21:in `block in <main>'
	from /usr/share/gems/gems/rake-13.2.1/lib/rake/rake_test_loader.rb:6:in `select'
	from /usr/share/gems/gems/rake-13.2.1/lib/rake/rake_test_loader.rb:6:in `<main>'
rake aborted!
Command failed with status (1): [ruby -w -I"lib:test" -Eascii-8bit:ascii-8bit /usr/share/gems/gems/rake-13.2.1/lib/rake/rake_test_loader.rb "test/plugin/test_compressable.rb" ]
/usr/share/gems/gems/rake-13.2.1/exe/rake:27:in `<top (required)>'
/usr/share/gems/gems/bundler-2.5.16/lib/bundler/cli/exec.rb:58:in `load'
/usr/share/gems/gems/bundler-2.5.16/lib/bundler/cli/exec.rb:58:in `kernel_load'
/usr/share/gems/gems/bundler-2.5.16/lib/bundler/cli/exec.rb:23:in `run'
/usr/share/gems/gems/bundler-2.5.16/lib/bundler/cli.rb:455:in `exec'
/usr/share/gems/gems/bundler-2.5.16/lib/bundler/vendor/thor/lib/thor/command.rb:28:in `run'
/usr/share/gems/gems/bundler-2.5.16/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/usr/share/gems/gems/bundler-2.5.16/lib/bundler/vendor/thor/lib/thor.rb:527:in `dispatch'
/usr/share/gems/gems/bundler-2.5.16/lib/bundler/cli.rb:35:in `dispatch'
/usr/share/gems/gems/bundler-2.5.16/lib/bundler/vendor/thor/lib/thor/base.rb:584:in `start'
/usr/share/gems/gems/bundler-2.5.16/lib/bundler/cli.rb:29:in `start'
/usr/share/gems/gems/bundler-2.5.16/exe/bundle:28:in `block in <top (required)>'
/usr/share/gems/gems/bundler-2.5.16/lib/bundler/friendly_errors.rb:117:in `with_friendly_errors'
/usr/share/gems/gems/bundler-2.5.16/exe/bundle:20:in `<top (required)>'
/usr/bin/bundle:25:in `load'
/usr/bin/bundle:25:in `<main>'
Tasks: TOP => test => base_test
(See full trace by running task with --trace)
```

@daipom i addressed your comments but couldnt test it locally due to some issue.
`yajl-ruby` is already installed in my system but some other issue

daipom commented on lib/fluent/plugin/in_forward.rb:
```diff
@@ -309,8 +309,12 @@ def on_message(msg, chunk_size, conn)
         # PackedForward
         option = msg[2]
         size = (option && option['size']) || 0
```
```suggestion
        option = msg[2] || {}
        size = option['size'] || 0
```

daipom commented on lib/fluent/plugin/in_forward.rb:
```diff
@@ -309,8 +309,12 @@ def on_message(msg, chunk_size, conn)
-        es_class = (option && option['compressed'] == 'gzip') ? Fluent::CompressedMessagePackEventStream : Fluent::MessagePackEventStream
-        es = es_class.new(entries, nil, size.to_i)
+
+        if option['compressed'] == 'gzip' && option['compressed'] != 'zstd'
```
```suggestion
        if option['compressed'] == 'gzip' || option['compressed'] == 'zstd'
```

daipom reviewed (changes requested):
Did not this code work?

https://github.com/fluent/fluentd/pull/4657#discussion_r1826526976

Athishpranav2003 commented:
I guess there is a small careless
Problem is local testing is not working at all. Could help me with that later

I will check after 18th
Got packed with exams now :(

daipom commented:
Hmm, these changes would be necessary, but looks like the error in your environment is caused by another reason.

https://github.com/fluent/fluentd/pull/4657#issuecomment-2453298305
> /usr/share/ruby/bundled_gems.rb:75:in `require': libruby.so.3.2: cannot open shared object file: No such file or directory - /usr/lib64/gems/ruby/yajl-ruby-1.4.3/yajl/yajl.so (LoadError)

daipom commented:
> I will check after 18th
Got packed with exams now :(

OK! I'm sorry my response was so slow, and it took so long.
Good luck with the exams!

Athishpranav2003 commented:
@daipom the gem issue got resolved after update,prolly some mismatched versions,

All tests are passing now

Athishpranav2003 commented:
@daipom some doubts here

in chuck open method for decompress we dont have any type identification.
how to identify the chunk type while open/read method is called

daipom commented:
> @daipom some doubts here
> 
> in chuck open method for decompress we dont have any type identification. how to identify the chunk type while open/read method is called
> 
> https://github.com/fluent/fluentd/blob/78a7972bfe6b421f08472701f04f00515ed24bee/lib/fluent/plugin/buffer/chunk.rb#L201-L226

I see!
The current implementation is based on the following assumption.

* It is a gzip format chunk = It extends `Decompressable` module

https://github.com/fluent/fluentd/blob/10f0f112eb112219bbe959a3e102102b0d429ce7/lib/fluent/plugin/buffer/chunk.rb#L51-L64

If we add a compression format, we need to change something here.
I discussed the direction with @Watson1978  and @kenhys.
We believe the following direction seems to be a good one.

* Create `Decompressable` module for each compression format
  * `GzipDecompressable` (Rename the current `Decompressable`)
  * `ZstdDecompressable`
* Switch modules to extend according to the compression format of the chunk

The current implementation of `Decompressable` looks a little weird to me.
I don't see the point of specifying the format in argument `compressed`
I don't want to change the signature for compatibility, but I believe a parameter like `raw` would be more correct...

Anyway, we only need to consider the possibility that `gzip` is specified for `compressed` for `GzipDecompressable`, and `zstd` for `ZstdDecompressable`.
(If a different compression format is specified to `compressed`, the chunk should ignore it and decompress itself.)

Athishpranav2003 commented:
Oh ok
Let me again go through and check it

But what is that parameter `compress` referring to?

Athishpranav2003 commented:
@daipom seems like `buffer` is working :)
Will go ahead and do other stuff

Athishpranav2003 commented:
@daipom Can you check this `out_file` once. Also how to go about UTs for it. Which all UTs should we write for zstd.

Athishpranav2003 commented:
@daipom can you check the PR now?

Athishpranav2003 commented:
@daipom added UTs as well.
PR seems to be complete now
let me know if i missed anything

daipom commented:
Sorry I have been busy this week :cry:
Thanks! I will check it out!

Athishpranav2003 commented:
hey @daipom sorry to ping again. 
would be good if i can wrap this up soon

daipom commented:
@Athishpranav2003 
> hey @daipom sorry to ping again. would be good if i can wrap this up soon

Sorry for my late response :cry:
I was busy releasing Fluentd v1.18 and fluent-package v5.2, sorry.
Now, those are done.
I will see this in a few days.

Athishpranav2003 commented:
Yah,
I was seeing ur contributions for a while. 
Its fine for me

daipom commented:
@Athishpranav2003 
Sorry for my late response.
Overall, it looks great! Thanks so much!

About `out_file` and `out_forward`, it would be better that we do not support setting a different compression format from the buffer.
There would be no particular benefit in allowing it.
Those plugins should raise `ConfigError` if a different compression format is set (except `:text`).

For example, for `out_forward`, we should check the format as follows.

```rb
        if @buffer.compress == :text
          @buffer.compress = @compress unless @compress == :text
        else
          if @compress == :text
            log.info "buffer is compressed.  If you also want to save the bandwidth of a network, Add `compress` configuration in <match>"
          elsif @compress != @buffer.compress
            raise Fluent::ConfigError, "You cannot specify different compression formats for Buffer (Buffer: #{@buffer.compress}, Self: #{@compress})"
          end
        end
```

As for implementation, there appears to be no other problems.
(There are some points where it would be nice to make some things simpler, but not essential.)

If you have any other concerns, please let me know.

We need to update the protocol before we merge this.
I will consider how to proceed it.

> To judge if we can merge this only with the support of the module and `in_forward`, we need to reach an agreement on updating [Forward Protocol Specification](https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1.5).
> 
> It would be necessary to update `CompressedPackedForward Mode`. By allowing `zstd` as the value for the key `compressed`, it would be possible to add Zstd support while keeping compatibility, but I'm not sure yet.

Athishpranav2003 commented:
@daipom i have addressed ur comments. Feel free to point out the nits as well.

daipom commented on lib/fluent/plugin/out_file.rb:
```diff
@@ -212,17 +213,27 @@ def write(chunk)
+                  elsif @buffer.compress == :gzip
+                    method(:write_from_compressed_chunk).curry.call(@compress_method)
+                  else
+                    raise Fluent::ConfigError, "You cannot specify different compression formats for Buffer (Buffer: #{@buffer.compress}, Self: #{@compress})"
```
Thanks!
Could you make this `raise` run on `configure`?
It would be better to check the ConfigError at startup as `out_forward`.

Athishpranav2003 replied:
@daipom made the changes accordingly

Athishpranav2003 commented:
@daipom 
we need to precisely update this section ryt? https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1.5#compressedpackedforward-mode

daipom commented on lib/fluent/plugin/out_file.rb:
```diff
@@ -184,6 +185,14 @@ def configure(conf)
         end
+
+        unless @buffer.compress == :text
+          if @compress == :text
+            log.info "buffer is compressed.  If you also want to save disk space, Add `compress` configuration in <match>"
+          elsif @buffer.compress != @compress_method
```
Thanks!

This log does not seem necessary.

> log.info "buffer is compressed. If you also want to save disk space, Add `compress` configuration in <match>"

Unlike `out_forward`, cases of compressing the buffer and not the output would be normal for `out_file`. 
(For `out_forward`, `in_forward` receives the output, so if the buffer is compressed, it is natural to compress the output too.)

Athishpranav2003 replied:
Force pushed these changes

daipom commented:
> @daipom we need to precisely update this section ryt? https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1.5#compressedpackedforward-mode

Yes!!

daipom commented:
> > @daipom we need to precisely update this section ryt? https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1.5#compressedpackedforward-mode
> 
> Yes!!

I have made an issue for this.

* #4758

Let's summarize the new protocol there, for now.

daipom commented:
> > > @daipom we need to precisely update this section ryt? https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1.5#compressedpackedforward-mode
> > 
> > 
> > Yes!!
> 
> I have made an issue for this.
> 
>     * [Update Forward Protocol Specification for zstd compression support #4758](https://github.com/fluent/fluentd/issues/4758)
> 
> 
> Let's summarize the new protocol there, for now.

@Athishpranav2003 
I have made a draft. Could you please check it?

https://github.com/fluent/fluentd/issues/4758#issuecomment-2572328791

daipom reviewed (approved):
LGTM! Thanks for this significant improvement!!
This should be merged into the next minor version v1.19.
Also, we need to wait #4758.

daipom reviewed (approved):
LGTM! Thanks so much!

In #4758, we now have a clear direction for protocol changes, so I merge this PR.
